component/publisher-client/src/main/java/org/wso2/am/analytics/publisher/client/MoesifClient.java

```diff
@@ -0,0 +1,206 @@
+                int statusCode = context.getResponse().getStatusCode();
+
+                if (statusCode >= 400 && statusCode < 500) {
+                    log.error("Event publishing failed. Moesif returned:", statusCode);
+                } else if (error != null) {
+                    log.error("Event publishing failed.", error);
+                } else {
+                    log.error("Event publishing failed. Retrying.");
+                    doRetry(orgId, builder);
+                }
+
+            }
+        };
+        try {
+            api.createEventAsync(buildEventResponse(event), callBack);
+        } catch (IOException e) {
+            log.error("Analytics event sending failed. Event will be dropped", e);
+        }
+
+    }
+
+    private EventModel buildEventResponse(Map<String, Object> data) throws IOException, MetricReportingException {
+        Gson gson = new Gson();
+        String jsonString = gson.toJson(data);
+        String reqBody = jsonString.replaceAll("[\r\n]", "");
+
+        //      Preprocessing data
+        final URL uri = new URL((String) data.get(Constants.DESTINATION));
+        final String hostName = uri.getHost();
+
+        final String userIP = (String) data.get(Constants.USER_IP);
+
+        Map<String, String> reqHeaders = new HashMap<String, String>();
+
+        reqHeaders.put("User-Agent",
+                (String) data.getOrDefault(Constants.USER_AGENT_HEADER, Constants.UNKNOWN_VALUE));
+        reqHeaders.put("Content-Type", Constants.MOESIF_CONTENT_TYPE_HEADER);
+        reqHeaders.put("Host", hostName);
+
+        Map<String, String> rspHeaders = new HashMap<String, String>();
+
+        rspHeaders.put("Vary", "Accept-Encoding");
+        rspHeaders.put("Pragma", "no-cache");
+        rspHeaders.put("Expires", "-1");
+        rspHeaders.put("Content-Type", "application/json; charset=utf-8");
+        rspHeaders.put("Cache-Control", "no-cache");
+
+
+        EventRequestModel eventReq = new EventRequestBuilder()
+                .time(new Date()) // See if you can parse request time stamp to date obj
+                .uri(uri.toString())
+                .verb((String) data.get(Constants.API_METHOD))
+                .apiVersion((String) data.get(Constants.API_VERSION))
+                .ipAddress(userIP)
+                .headers(reqHeaders)
+                .body(reqBody)
+                .build();
+
+
+        EventResponseModel eventRsp = new EventResponseBuilder()
+                .time(new Date(System.currentTimeMillis() + 1000))
+                .status((int) data.get(Constants.TARGET_RESPONSE_CODE))
+                .headers(rspHeaders)
+                .build();
+
+        EventModel eventModel = new EventBuilder()
+                .request(eventReq)
+                .response(eventRsp)
+                .userId((String) data.get("userName"))
+                .companyId((String) data.get(Constants.ORGANIZATION_ID))
+                .build();
+
+        return eventModel;
```

Comment:
no need eventModel. can return the new instance.

Reply:
Didn't get it Rukshan.
Can you pls bit explain?